simSPI/multislice.py

```diff
@@ -12,18 +13,22 @@ def apply_complex_ctf_to_exit_wave(exit_wave_f, complex_ctf):
 
     Parameters
     ----------
-    exit_wave_f : numpy.ndarray, shape (N,N)
+    exit_wave_f : numpy.ndarray, shape (n_pixels,n_pixels)
         Exit wave array.
-    complex_ctf : numpy.ndarray, shape (N,N)
+    complex_ctf : numpy.ndarray, shape (n_pixels,n_pixels)
         Complex ctf array.
 
     Returns
     -------
-    i0 : numpy.ndarray, shape (N,N)
+    i0 : numpy.ndarray, shape (n_pixels,n_pixels)
         Exit wave ctf convolution (in real space).
     """
-    i0 = np.abs(fourier.do_ifft(exit_wave_f * complex_ctf, dim=2, only_real=False))
-    return i0
+    n_pixels = exit_wave_f.shape[0]
+    np_to_torch = torch.tensor(exit_wave_f * complex_ctf).reshape(
+        1, 1, n_pixels, n_pixels
+    )
+    i0 = torch.abs(transforms.fourier_to_primal_2D(np_to_torch))
+    return i0.detach().numpy().reshape(n_pixels, n_pixels)
```

Comment:
An alternative would be to add the numpy FFT in compSPI/compSPI. Making it a backend option

Reply:
How should the backend be recognized? I don't have experience with something like this. 

The things that come to mind for me seem hacky. Like in `transforms.py`

```
def fourier_to_primal_2D(array):
  if isinstance(array,numpy.ndarray):
    # fourier fft
  elif torch.is_tensor(array):
    # torch fft
```
    
But I'd like more direction on how to make things general before coding something up.

Reply:
@ninamiolane do you have any suggestions for how to have things backend agnostic?

Reply:
Sure, either using eagerpy or having a backend repository (e.g. like in geomstats https://github.com/geomstats/geomstats/tree/master/geomstats/_backend).

But it's a lot of work...